rudder-core/src/main/scala/com/normation/rudder/services/reports/ExecutionBatch.scala

```diff
@@ -968,326 +1023,193 @@ object ExecutionBatch extends Loggable {
-        }
-    }
-
-    // Regroup all None value into None component Value
-    // There should only one None report per component
-    val (noneValue, noneReports) = {
-      if(valueKind.none.isEmpty) {
-      (None, Seq())
-      } else {
-        val reports = filteredReports.filter(r => r.keyValue == DEFAULT_COMPONENT_KEY)
-        ( Some(buildComponentValueStatus(
-              DEFAULT_COMPONENT_KEY
-            , reports
-            , unexpectedStatusReports.isEmpty
-            , valueKind.none.size
-            , noAnswerType
-            , policyMode
-          ))
-        , reports
-        )
-      }
-    }
-
-    // Create a simpleValue for each unexpandedValue, with all reports matching expanded values
-    val (simpleValues, simpleReports) = {
-      val result = for {
-        (unexpandedValue, values) <- valueKind.simple
-      } yield {
-        // we find the reports matching these values, but take
-        // at most the number of values if there is also cfevars
-        // because the remaining values may be for cfengine vars
-        val reports = {
-          val possible = filteredReports.filter(r => values.contains(r.keyValue))
-          if(valueKind.cfeVar.size > 0) {
-            possible.take(values.size)
-          } else {
-            possible
+    def recPairReports(reports: List[ResultReports], freeValues: List[Value], pairedValues: List[Value], unexpected: List[ResultReports], mode: UnexpectedReportInterpretation): (List[Value], List[ResultReports]) = {
+      // utility function: given a list of Values and one report, try to find a value whose pattern matches reports component value.
+      // the first matching value is used (they must be sorted by specificity).
+      // Returns (list of unmatched values, Option[matchedValue])
+      // We also have two special case to deal with:
+      // - in some case, we want to ignore a report totally (ie: it must not change compliance). This is typically for
+      //   what we want for duplicated reports. We need to log the report erasure as it won't be displayed in UI (by
+      //   design). The log level should be "info" and not more because it was chosen by configuration to ignore them.
+      // - in some case, we want to accept more reports than originally expected. Then, we must update cardinality to
+      //   trace that decision. It's typically what we want to do for
+      def findMatchinValue(
+          report              : ResultReports
+        , values              : List[Value]
+        , dropDuplicated      : (Value, ResultReports) => Boolean
+        , incrementCardinality: (Value, ResultReports) => Boolean
+      ): (List[Value], Option[Value]) = {
+        values.foldLeft((Nil:List[Value]), Option.empty[Value]) { case ((stack, found), value) =>
+          found match {
+            case Some(x) => (value :: stack, Some(x))
+            case None    =>
+              if(value.pattern.matcher(report.keyValue).matches()) {
```

Comment:
i'm probably missing something something there - we are only only evaluating value if they match a pattern, but most values don't have a pattern - reevaluating twice a pattern (once when we detect if there is a variable, then if it match) may be pretty costly - even more as we have a quadratic complextity

Reply:
ok